### DISCUSSION TITLE
Creating and Setting up Tenant via workflow does not work with liquid expression for admin email.

### DISCUSSION THREAD

**Discussion by Mordsspass** [OrchardCMS/OrchardCore#11828] [Q&A]
We first had a Workflow, in which the admin E-mail of a new tenant is constant throguhout workflows .
![image](https://user-images.githubusercontent.com/66907796/172589346-77ce2d98-8a8c-49f3-9501-c3104b9bc9c3.png)
(Image of The Setup Tenant Task )
Our Workflow was for a Customer to fill out a form and send the data. -> Customer needs to confirm -> someone else needs to approve -> Tenant is automatically created. (here we had a default Account which was created as SuperUser and the Customer Data was used to create a User in this new Tenant.) but we want to change it so that the customer is the superadmin. 

We save the Form data for the duration of the run in Properties :
![image](https://user-images.githubusercontent.com/66907796/172595065-502d2948-ac61-42bb-b8ac-a108dd341880.png)

Now to the Problem: When we use a Liquid Expression like 
![image](https://user-images.githubusercontent.com/66907796/172590128-5cf4b244-241f-4ba8-9123-693fdd4c1ccb.png)
It fails the Task and shows Invalid Email Exception. The Property has a value which is an accpetable Email- Address.
(firstname.lastname@domain.com)

We were thinking it was an Encoding which happens. Is there an issue with the way the Property is saved?
Hope someone can help. 

Appended Strings as Liquid Expression does work. `{{"Test" | append: "@test.com"}}` but to get the value of the workflow apparently does not.

Is it possible that the value of the Email which is safed in the Database is not the actual Value the Expressions stands for but the Expression itself? ( We tested with `{{Workflow.Properties["Email"] | default: "test.test@test.com"}}` and it worked for the Setup but it showed this exact string as the User Email. )

**Comment by ns8482e** [comment]
It seems it’s null or empty for  `Workflow.Properties["Email"]` did you verify if that returns the email you intend to use?

**Reply by Mordsspass** [reply]
Yes did indeed test it. Both withing code itself and within workflow. ( Made a Notify Task -> showed the correct email )
The Workflow just failed the Setup Tenant task with Microsoft.AspNetCore.Identity.UserManager|WARN|User validation failed: InvalidEmail.
Is it possible that liquid needs a specific encoding to work with Email adresses?

**Reply by jtkech** [reply]
Can you show us your code, e.g. how you get your `Form`. Are you sure about your key names, I assume yes ;)

Hmm, `Request.Form[somekey]` doesn't return a string but a `StringValues` (may contain only one string) that can be easily converted to a string, don't remember if we can just do a cast `(string)Form[...]`, at least you can try a `ToString()`.

**Reply by Mordsspass** [reply]
Thank you for your time. We have a custom razor page which triggers the Workflow.

```
<form class="p-4" method="post" action="@("/workflows/Invoke?token="+Model.token)" enctype="application/x-www-form-urlencoded">

        @Html.AntiForgeryToken()<br>
        <div class="card p-3">
            <h4>Account Details</h4>
            <hr />
            <div asp-validation-summary="All" class="text-danger"></div>
            <div></div>
            <div class="form-group">
                <label for="CompanyName">Company Name</label>
                <input id="CompanyName" name="companyName" placeholder="Company Name" type="text" v-model="companyName" required
                class="form-control @(ModelState.GetValidationState("companyName")==
                Microsoft.AspNetCore.Mvc.ModelBinding.ModelValidationState.Invalid?"is-invalid":"")">
            </div>
            <div class="form-group">
                <label for="tenantName">Tenant Name</label>
                <input id="tenantName" name="tenantName" placeholder="Tenant Name" v-model="tenantName" type="text" required
                class="form-control @(ModelState.GetValidationState("tenantName")==
                Microsoft.AspNetCore.Mvc.ModelBinding.ModelValidationState.Invalid?"is-invalid":"")">
            </div>
        </div>
        <div class="card p-3 mt-2">
            <h4>Account Root User Details</h4>
            <div class="form-group">
                <label for="userName">Username</label>
                <input id="userName" name="userName" v-model="userName" placeholder="Admin" type="text" required class="form-control @(ModelState.GetValidationState("userName")==Microsoft.AspNetCore.Mvc.ModelBinding.ModelValidationState.Invalid?"is-invalid":"")" />
            </div>
            <div class="form-group">
                <label for="userEmail">E-Mail</label>
                <input v-bind:class="" id="userEmail" name="userEmail" v-model="email"
                placeholder="Admin@yourCompany.com" type="email" required 
                class="form-control @(ModelState.GetValidationState("userEmail")==
                Microsoft.AspNetCore.Mvc.ModelBinding.ModelValidationState.Invalid?"is-invalid":"")" />
            </div>
            <div class="form-group">
                <label for="Password">Password</label>
                <input id="Password" name="Password" v-model="password" type="password"
                       data-bs-placement="top" data-bs-toggle="tooltip"
                       title="Password needs to have 8 characters and at least 1 upper case, 1 lower case and 1 special character"
                       required class="form-control @(ModelState.GetValidationState("Password")==
                       Microsoft.AspNetCore.Mvc.ModelBinding.ModelValidationState.Invalid?"is-invalid":"")">
            </div>
            <div class="form-group">
                <label for="ConfirmPassword">Confirm Password</label>
                <input id="ConfirmPassword" name="ConfirmPassword" v-model="confirmPassword" type="password" required 
                class="form-control @(ModelState.GetValidationState("ConfirmPassword")==
                Microsoft.AspNetCore.Mvc.ModelBinding.ModelValidationState.Invalid?"is-invalid":"")">
            </div>
```

The Page with the Form.
`var Form = context.Request.Form;` To get the FormData from the incoming Request. ( for the Code in the discussion.

They get validated in the backend and than the values are safed for use after pauses ( waiting for User actions). This is done with the Code above ( workflowContext. ..... ) 
Casting to string/ calling the ToString Method does not solve the problem.

**Comment by jtkech** [comment]
We use `MimeKit` for email validation, I tried to setup a tenant with `admin@admin.com`, it works.

Hmm `v-model="email"`, do you mean `v-model="userEmail"`?

**Reply by Mordsspass** [reply]
We also tried to manually type the email. the email itselve was never the problem. It somehow did not interpret the liquid as a valid Email.
To make sure we even hard coded a Workflow.Properties["Email"] to a valid email for Example : "firstname.lastname@domain.com". It still said it was an invalid email.
